upup/pkg/fi/cloudup/urls.go

```diff
@@ -49,42 +51,85 @@ func BaseUrl() string {
+func NodeUpLocation(assetsBuilder *assets.AssetBuilder) (string, *hashing.Hash, error) {
+	if nodeUpLocation != "" && nodeUpHash != nil {
+		glog.V(10).Infof("returning saved nodeup values %q, %q", nodeUpLocation, nodeUpHash)
+		return nodeUpLocation, nodeUpHash, nil
 	}
-	nodeUpLocation = os.Getenv("NODEUP_URL")
-	if nodeUpLocation == "" {
-		nodeUpLocation = BaseUrl() + "linux/amd64/nodeup"
-		glog.V(4).Infof("Using default nodeup location: %q", nodeUpLocation)
-	} else {
+	var err error
+	n := os.Getenv("NODEUP_URL")
+	var h *hashing.Hash
+	if n != "" {
 		glog.Warningf("Using nodeup location from NODEUP_URL env var: %q", nodeUpLocation)
+	} else {
+		n, h, err = FileUrl("linux/amd64/nodeup", assetsBuilder)
+		if err != nil {
+			return "", nil, err
+		}
+		glog.V(4).Infof("Using default nodeup location: %q", n)
 	}
-	return nodeUpLocation
+
+	nodeUpLocation = n
+	nodeUpHash = h
+
+	glog.V(10).Infof("returning values %q, %q", nodeUpLocation, nodeUpHash)
+	return n, h, nil
 }
 
 // protokubeImageSource caches the ProtokubeImageSource value
 var protokubeImageSource string
+var protokubeImageSHA *hashing.Hash
+
+// TODO FIXME make this a container when hosted assets
 
 // ProtokubeImageSource returns the source for the docker image for protokube.
 // Either a docker name (e.g. gcr.io/protokube:1.4), or a URL (https://...) in which case we download
 // the contents of the url and docker load it
-func ProtokubeImageSource() string {
-	if protokubeImageSource != "" {
+func ProtokubeImageSource(assetsBuilder *assets.AssetBuilder) (string, *hashing.Hash, error) {
+	if protokubeImageSource != "" && protokubeImageSHA != nil {
 		// Avoid repeated logging
-		return protokubeImageSource
+		glog.V(4).Infof("Using returning saved protokube location: %q", protokubeImageSource)
+		return protokubeImageSource, protokubeImageSHA, nil
 	}
 	protokubeImageSource = os.Getenv("PROTOKUBE_IMAGE")
 	if protokubeImageSource == "" {
-		protokubeImageSource = BaseUrl() + "images/protokube.tar.gz"
+		p, h, err := FileUrl("images/protokube.tar.gz", assetsBuilder)
+		if err != nil {
+			return "", nil, err
+		}
+		protokubeImageSHA = h
+		protokubeImageSource = p
 		glog.V(4).Infof("Using default protokube location: %q", protokubeImageSource)
+		return p, h, nil
 	} else {
+		if protokubeImageSHA == nil {
+			// FIXME get SHA
+		}
 		glog.Warningf("Using protokube location from PROTOKUBE_IMAGE env var: %q", protokubeImageSource)
 	}
-	return protokubeImageSource
+
+	return protokubeImageSource, protokubeImageSHA, nil
+
+}
+
+// BaseUrl returns the base url for the distribution of kops - in particular for nodeup & docker images
+func FileUrl(file string, assetBuilder *assets.AssetBuilder) (string, *hashing.Hash, error) {
+
+	b := os.Getenv("KOPS_BASE_URL")
+	if b != "" {
+		glog.Warningf("Using base url from KOPS_BASE_URL env var: %q", baseUrl)
+	} else {
+		b = kopsUrl
+	}
+
+	fileUrl, hash, err := assetBuilder.RemapFileAndSHA(b+file, b+file+".sha1")
```

Comment:
I think you probably want to use `filepath.Join` here

Reply:
Yes, but that is not why e2e bombed.

Reply:
not certain it is there, but yes...

Reply:
Little more complicated to do a url ... 

```
	baseUrl, err := url.Parse(baseUrlString)
        baseUrl.Path = path.Join(u.baseUrl, file)
        s := u.String()
```